File: ios/Classes/CBManager.swift
//
//  CBManager.swift
//  Runner
//
//  Created by Luca Christille on 14/08/18.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift

class CBManager {
    // static let instance = CBManager();
    private var mDatabase : Dictionary<String, Database> = Dictionary();
    private var mReplConfig : ReplicatorConfiguration! = nil;
    private var mReplicator : Replicator! = nil;
    private var defaultDatabase = "defaultDatabase";
    
    init() {}
    
    func getDatabase() -> Database? {
        if let result = mDatabase[defaultDatabase] {
            return result;
        } else {
            return nil;
        }
    }
    
    func getDatabase(name : String) -> Database? {
        if let result = mDatabase[name] {
            return result;
        } else {
            return nil;
        }
    }
    
    func saveDocument(map: Dictionary<String, Any>) throws -> String? {
        let mutableDocument: MutableDocument = MutableDocument(data: map);
        try mDatabase[defaultDatabase]?.saveDocument(mutableDocument)
        return mutableDocument.id;
    }
    
    func saveDocumentWithId(id : String, map: Dictionary<String, Any>) throws -> String? {
        if let defaultDb: Database = getDatabase() {
            if let doc: Document = defaultDb.document(withID: id) {
                let mutableMap = (map as NSDictionary).mutableCopy() as! NSMutableDictionary;
                for key in doc.keys{
                    if(key == "_attachments"){
                        mutableMap[key] = doc.value(forKey: key)
                    }
                    if let blob = doc.blob(forKey: key){
                        mutableMap[key] = blob
                    }
                }
                let resultMap = removeAttachmentsPath(object: mutableMap as NSDictionary)
                let mutableDocument: MutableDocument = MutableDocument(id: id, data: mutableMap as? Dictionary<String, Any>)
                try defaultDb.saveDocument(mutableDocument)
                return mutableDocument.id
            }
        }
        return id
    }
    
    func removeAttachmentsPath(object: AnyObject) -> AnyObject {
        if let dict = object as? NSDictionary{
            let d:NSMutableDictionary = dict.mutableCopy() as! NSMutableDictionary
            if(d.value(forKey: "contentType") != nil){
                d["path"] = nil
            }
            for (key, value) in d {
                let val =  removeAttachmentsPath(object:value as AnyObject)
                d[key] = val;
            }
        }else if let array = object as? NSArray{
            let a =  array.map{(value)->AnyObject in
                return removeAttachmentsPath(object:value as AnyObject)
                } as AnyObject
            return a
        }
        return object;
        
    }
    
    func addAttachmentsPath(attarchments:NSDictionary, object:AnyObject,  key:String?) -> AnyObject {
        if let dict = object as? NSDictionary {
            let d:NSMutableDictionary = dict.mutableCopy() as! NSMutableDictionary
            if let attachKey = d["file"] as! String?{
                if let attachPath = attarchments[attachKey] {
                    d["path"] = attachPath
                }
            }
           
            for (key, value) in dict {
                let val = addAttachmentsPath(attarchments: attarchments, object: value as AnyObject, key: key as? String)
                d[key] = val
            }
            return d as AnyObject
        }
        else if let array = object as? NSArray {
            let a =  array.map{(value)->AnyObject in
                return addAttachmentsPath(attarchments: attarchments, object: value as AnyObject, key: nil)
                } as AnyObject
            return a
        }
        return object;
    }
    
    
    
    func processDoc (document: Document) -> NSDictionary? {
        let dictionary: NSDictionary = document.toDictionary() as NSDictionary
        var retrievedDocument: NSMutableDictionary = dictionary.mutableCopy() as! NSMutableDictionary
        let attachmentsPath = NSMutableDictionary()
        for (key, _) in retrievedDocument {
            if let blob = document.blob(forKey: (key as! String)){
                 attachmentsPath[key] = blob.filePath
            }
        }
        for (key, _) in attachmentsPath {
            retrievedDocument.removeObject(forKey: key as! String)
        }
        if let attcahments = retrievedDocument.value(forKey: "_attachments") as? NSDictionary {
            
            for (key, value) in attcahments {
                if let blob = value as? Blob {
                    attachmentsPath[key] = blob.filePath
                }
            }

            retrievedDocument.removeObject(forKey: "_attachments")

        }

        if let d = addAttachmentsPath(attarchments: attachmentsPath, object: retrievedDocument, key: nil) as? NSDictionary {
            retrievedDocument = d.mutableCopy() as! NSMutableDictionary
        }
        
        return retrievedDocument
    }
    
    func getDocumentWithId(id : String) -> NSDictionary? {
        let resultMap: NSMutableDictionary = NSMutableDictionary()
        resultMap["id"] = id
        resultMap["doc"] = NSDictionary()
        if let defaultDb: Database = getDatabase() {
            if let document: Document = defaultDb.document(withID: id) {
                if let d = processDoc(document: document) {
                    resultMap["id"] = document.id
                    resultMap["doc"] = d
                }
            } else {
                resultMap["id"] = id
                resultMap["doc"] = NSDictionary.init()
            }
        }
        return NSDictionary.init(dictionary: resultMap)
    }

    func getDocumentsWith(key : String, value: String) -> NSDictionary? {
        let resultMap: NSMutableDictionary = NSMutableDictionary();
        if let defaultDb: Database = getDatabase() {
            let query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb))
                .where(Expression.property(key).equalTo(Expression.string(value)))
            do {
                let result = try query.execute()
                let docs = result.allResults().map{(result)->NSDictionary in
                    let ret = NSMutableDictionary();
                    if let id = result.string(forKey: "id") {
                        if let doc:Document = defaultDb.document(withID: id) {
                        ret["doc"] = processDoc(document: doc)
                        ret["id"] = id
                        }
                    } 
                    return ret;
                    };
                resultMap["docs"] = docs;
            } catch {
                resultMap["docs"] = nil;
            }
        }
        return resultMap;
    }


    func getAllDocuments() -> NSDictionary {
        let resultMap: NSMutableDictionary = NSMutableDictionary();
        if let defaultDb: Database = getDatabase() {
            let query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.database(defaultDb))
            do {
                let result = try query.execute()
                let docs = result.allResults().map{(result)->NSDictionary in
                    let ret = NSMutableDictionary();
                    if let id = result.string(forKey: "id") {
                        if let doc:Document = defaultDb.document(withID: id) {
                            ret["doc"] = processDoc(document: doc)
                            ret["id"] = id
                        }
                    }
                   
                    return ret;
                    };
                resultMap["docs"] = docs;
            } catch {
                resultMap["docs"] = nil;
            }
        }
        return resultMap;
    }
    
    func addAttachment(docId:String, contentType:String, filePath:String) throws -> String? {
        if let defaultDb: Database = getDatabase() {
            if let doc: MutableDocument = defaultDb.document(withID: docId)?.toMutable() {
                let key = String((0..<5).map{ _ in "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".randomElement()! })//String(Int.random(in: 0 ..< 99999)) //TODO random String
                let blob = try Blob(contentType: contentType, fileURL: URL.init(fileURLWithPath: filePath))
                doc.setBlob(blob, forKey: key)
                try defaultDb.saveDocument(doc)
                return key;
            }
        }
        return nil;
    }
    
    func removeAttachment(docId:String, key:String) throws {
        if let defaultDb: Database = getDatabase() {
            if let doc: MutableDocument = defaultDb.document(withID: docId)?.toMutable() {
                doc.setBlob(nil, forKey: key)
                try defaultDb.saveDocument(doc)
            }
        }
    }
    
    func purgeDocument(docId: String) -> Bool {
        if let defaultDb: Database = getDatabase() {
            if let doc: Document = defaultDb.document(withID: docId) {
                try! defaultDb.purgeDocument(doc);
                return true;
            }
        }
        return false;
    }

    func deleteDocument(docId: String) -> Bool {
        if let defaultDb: Database = getDatabase() {
            if let doc: Document = defaultDb.document(withID: docId) {
                try! defaultDb.deleteDocument(doc);
                return true;
            }
        }
        return false;
    }

    func initDatabaseWithName(name: String){
        if mDatabase.keys.contains(name) {
            defaultDatabase = name
        } else {
            do {
                let newDatabase = try Database(name: name)
                // Database.setLogLevel(level: LogLevel.verbose, domain: LogDomain.replicator)
                mDatabase[name] = newDatabase
                defaultDatabase = name
            } catch {
                print("Error initializing new database")
            }
        }
    }

     func delete(){
        if let defaultDb: Database = getDatabase() {
             try! defaultDb.delete();
             mDatabase.removeAll();
        }
     }
    
    func setReplicatorEndpoint(endpoint: String) {
        let targetEndpoint = URLEndpoint(url: URL(string: endpoint)!)
        mReplConfig = ReplicatorConfiguration(database: getDatabase()!, target: targetEndpoint)
    }
    
    func setReplicatorType(type: String) -> String {
        var settedType: ReplicatorType = ReplicatorType.pull
        if (type == "PUSH") {
            settedType = .push
        } else if (type == "PULL") {
            settedType = .pull
        } else if (type == "PUSH_AND_PULL") {
            settedType = .pushAndPull
        }
        mReplConfig?.replicatorType = settedType
        switch(mReplConfig?.replicatorType.rawValue) {
        case (0):
            return "PUSH_AND_PULL"
        case (1):
            return "PUSH"
        case(2):
            return "PULL"
        default:
            return ""
        }
    }
    
    func setReplicatorAuthentication(auth: [String:String]) -> String {
        if let username = auth["username"], let password = auth["password"] {
            mReplConfig?.authenticator = BasicAuthenticator(username: username, password: password)
        }
        return mReplConfig.authenticator.debugDescription
    }
    
    func setReplicatorSessionAuthentication(sessionID: String?) {
        if ((sessionID) != nil) {
            if ((mReplConfig) != nil) {
            mReplConfig.authenticator = SessionAuthenticator(sessionID: sessionID!)
            }
        }
    }
    
    func setReplicatorContinuous(isContinuous: Bool) -> Bool {
        if ((mReplConfig) != nil) {
            mReplConfig?.continuous = isContinuous
            return mReplConfig!.continuous
        }
        return false
    }
    
    func initReplicator() {
        mReplicator = Replicator(config: mReplConfig)
    }
    
    func startReplication() {
        if ((mReplicator) != nil) {
            mReplicator.start()
        }
    }
    
    func stopReplication() {
        if ((mReplicator) != nil) {
            mReplicator.stop()
            mReplicator = nil
        }
    }
    
    func getReplicator() -> Replicator {
        return mReplicator
    }
}

File: android/src/main/java/it/oltrenuovefrontiere/fluttercouch/CBManager.java
package it.oltrenuovefrontiere.fluttercouch;

import android.util.Log;

import com.couchbase.lite.BasicAuthenticator;
import com.couchbase.lite.Blob;
import com.couchbase.lite.CouchbaseLite;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseChangeListener;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.Document;
import com.couchbase.lite.Endpoint;
import com.couchbase.lite.Expression;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.Meta;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Replicator;
import com.couchbase.lite.ReplicatorChangeListener;
import com.couchbase.lite.ReplicatorConfiguration;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.SessionAuthenticator;
import com.couchbase.lite.URLEndpoint;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class CBManager {

    private Database mDatabase;
    private ReplicatorConfiguration mReplicatorConfig;
    private Replicator mReplicator;
    private String attachPath;

    public void initDatabaseWithName(String _name) throws CouchbaseLiteException {
        CouchbaseLite.init(FluttercouchPlugin.context);
        DatabaseConfiguration config = new DatabaseConfiguration();
        if (mDatabase == null) {
            mDatabase = new Database(_name, config);
        }
        attachPath = mDatabase.getPath();
    }

    public void close() throws CouchbaseLiteException {
        ensureInitialized();
        mDatabase.close();
        mDatabase = null;
    }

    public void delete() throws CouchbaseLiteException {
        ensureInitialized();
        mDatabase.delete();
        mDatabase = null;
    }

    public String saveDocument(Map<String, Object> _map) throws CouchbaseLiteException {
        ensureInitialized();
        MutableDocument mutableDoc = new MutableDocument(_map);
        mDatabase.save(mutableDoc);
        return mutableDoc.getId();
    }

    public ListenerToken addDatabaseChangeListener(DatabaseChangeListener listener) {
        ensureInitialized();
        return mDatabase.addChangeListener(listener);
    }

    public void removeDatabaseChangeListener(ListenerToken token) {
        ensureInitialized();
        mDatabase.removeChangeListener(token);
    }


    public ListenerToken addReplicationChangeListener(ReplicatorChangeListener listener) {
        ensureInitialized();
        return mReplicator.addChangeListener(listener);
    }

    public void removeReplicationChangeListener(ListenerToken token) {
        ensureInitialized();
        mReplicator.removeChangeListener(token);
    }

    public String saveDocumentWithId(String _id, Map<String, Object> _map) throws CouchbaseLiteException {
        ensureInitialized();
        Document document = mDatabase.getDocument(_id);
        for (String key : document.getKeys()) {
            if (key.equals("_attachments")) {
                _map.put(key, document.getValue(key));
            }
            Blob b = document.getBlob(key);
            if (b != null) _map.put(key, b);
        }
        removeAttachmentsPath(_map);
        MutableDocument mutableDoc = new MutableDocument(_id, _map);
        mDatabase.save(mutableDoc);
        return mutableDoc.getId();
    }

    public Map<String, Object> getDocumentWithId(String _id) throws CouchbaseLiteException {
        ensureInitialized();
        Map<String, Object> resultMap = new HashMap<>();
        try {
            Document document = mDatabase.getDocument(_id);
            if (document != null) {
                resultMap.put("doc", processDoc(document));
                resultMap.put("id", document.getId());
            } else {
                resultMap.put("doc", null);
                resultMap.put("id", _id);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return resultMap;
    }

    public Map<String, Object> getDocumentsWith(String key, String value) throws CouchbaseLiteException {
        ensureInitialized();
        HashMap<String, Object> resultMap = new HashMap<String, Object>();
        Query query = QueryBuilder.select(SelectResult.expression(Meta.id))
                .from(DataSource.database(mDatabase))
                .where(Expression.property(key).equalTo(Expression.string(value)));
        try {
            ResultSet result = query.execute();
            ArrayList docs = new ArrayList();
            for (Result res : result.allResults()) {
                HashMap<String, Object> ret = new HashMap<String, Object>();
                ret.put("doc", processDoc(mDatabase.getDocument(res.getString("id"))));
                ret.put("id", res.getString("id"));
                docs.add(ret);
            }
            resultMap.put("docs", docs);
        } catch (Exception e) {
            e.printStackTrace();
            resultMap.put("docs", null);
        }
        return resultMap;
    }

    public Map<String, Object> getAllDocuments() throws CouchbaseLiteException {
        ensureInitialized();
        HashMap<String, Object> resultMap = new HashMap<String, Object>();
        Query query = QueryBuilder.select(SelectResult.expression(Meta.id))
                .from(DataSource.database(mDatabase));
        try {
            ResultSet result = query.execute();
            ArrayList docs = new ArrayList();
            for (Result res : result.allResults()) {
                HashMap<String, Object> ret = new HashMap<String, Object>();
                ret.put("doc", processDoc(mDatabase.getDocument(res.getString("id"))));
                ret.put("id", res.getString("id"));
                docs.add(ret);
            }
            resultMap.put("docs", docs);
        } catch (Exception e) {
            e.printStackTrace();
            resultMap.put("docs", null);
        }
        return resultMap;
    }

    public String addAttachment(String _id, String contentType, InputStream is) throws CouchbaseLiteException, IOException {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            ensureInitialized();
            MutableDocument document = mDatabase.getDocument(_id).toMutable();
            String key = new RandomString(5, new Random()).nextString();
            IOUtils.inputStreamToOutputStream(is, os);
            Blob b = new Blob(contentType, os.toByteArray());
            document.setBlob(key, b);
            mDatabase.save(document);
            return key;
        } finally {
            IOUtils.closeSafe(is);
            IOUtils.closeSafe(os);
        }

    }

    public void removeAttachment(String _id, String key) throws CouchbaseLiteException {
        ensureInitialized();
        MutableDocument document = mDatabase.getDocument(_id).toMutable();
        document.setBlob(key, null);
        mDatabase.save(document);
    }

    public void purgeDocument(String _id) throws CouchbaseLiteException {
        ensureInitialized();
        Document document = mDatabase.getDocument(_id);
        if (document != null) {
            mDatabase.purge(document);
        }
    }

    public void deleteDocument(String _id) throws CouchbaseLiteException {
        ensureInitialized();
        Document document = mDatabase.getDocument(_id);
        if (document != null) {
            mDatabase.delete(document);
        }
    }


    public String setReplicatorEndpoint(String _endpoint) throws URISyntaxException {
        ensureInitialized();
        Endpoint targetEndpoint = new URLEndpoint(new URI(_endpoint));
        mReplicatorConfig = new ReplicatorConfiguration(mDatabase, targetEndpoint);
        return mReplicatorConfig.getTarget().toString();
    }

    public String setReplicatorType(String _type) throws CouchbaseLiteException {
        ReplicatorConfiguration.ReplicatorType settedType = ReplicatorConfiguration.ReplicatorType.PULL;
        if (_type.equals("PUSH")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PUSH;
        } else if (_type.equals("PULL")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PULL;
        } else if (_type.equals("PUSH_AND_PULL")) {
            settedType = ReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL;
        }
        mReplicatorConfig.setReplicatorType(settedType);
        return settedType.toString();
    }

    public String setReplicatorBasicAuthentication(Map<String, String> _auth) throws Exception {
        if (_auth.containsKey("username") && _auth.containsKey("password")) {
            mReplicatorConfig.setAuthenticator(new BasicAuthenticator(_auth.get("username"), _auth.get("password").toCharArray()));
        } else {
            throw new Exception();
        }
        return mReplicatorConfig.getAuthenticator().toString();
    }

    public String setReplicatorSessionAuthentication(String sessionID) throws Exception {
        if (sessionID != null) {
            mReplicatorConfig.setAuthenticator(new SessionAuthenticator(sessionID));
        } else {
            throw new Exception();
        }
        return mReplicatorConfig.getAuthenticator().toString();
    }

    public boolean setReplicatorContinuous(boolean _continuous) {
        mReplicatorConfig.setContinuous(_continuous);
        return mReplicatorConfig.isContinuous();
    }

    public void initReplicator() {
        mReplicator = new Replicator(mReplicatorConfig);
    }

    public void startReplicator() {
        mReplicator.start();
    }

    public void stopReplicator() {
        mReplicator.stop();
        mReplicator = null;
    }

    public Replicator getReplicator() {
        return mReplicator;
    }

    private String getBlobPath(Blob blob) {
        return  attachPath +"Attachments/" + blob.digest().substring(5).replace("/", "_") + ".blob";
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> processDoc(Document document) {
        Map<String, Object> retrievedDocument = new HashMap<>(document.toMap());
        Map<String, Object> attachmentsPath = new HashMap<>();

        for (String key : retrievedDocument.keySet()) {
            Blob b = document.getBlob(key);
            if (b != null) {
                attachmentsPath.put(key, getBlobPath(b));
            }
        }

        for (String key : attachmentsPath.keySet()) {
            retrievedDocument.remove(key);
        }
        Map<String, Object> attachments = (Map<String, Object>) retrievedDocument.get("_attachments");
        if (attachments != null) {
            for (String key : attachments.keySet()) {
                Object b = attachments.get(key);

                if (b instanceof Blob) {
                    attachmentsPath.put(key, getBlobPath(((Blob) b)));
                }
            }
            retrievedDocument.remove("_attachments");
        }

        if (attachmentsPath.size() > 0) {
            retrievedDocument = (Map<String, Object>) addAttachmentsPath(attachmentsPath, retrievedDocument, null);
        }
        return retrievedDocument;
    }


    private void removeAttachmentsPath(final Object object) {
        if (object instanceof Map) {
            Map<String, Object> map = ((Map<String, Object>) object);
            if (map.containsKey("contentType")) {
                map.remove("path");
            }
            for (String key : map.keySet()) {
                removeAttachmentsPath(map.get(key));
            }
        } else if (object instanceof List) {
            List<Object> list = (List<Object>) object;
            for (Object o : list) {
                removeAttachmentsPath(o);
            }
        }
    }

    @SuppressWarnings("unchecked")
    private Object addAttachmentsPath(final Map<String, Object> attachments, final Object object, String key) {
        if (object instanceof Map) {
            Map<String, Object> dict = ((Map<String, Object>) object);
            Map<String, Object> d = new HashMap<>(dict);
            Object attachPath = attachments.get((String) d.get("file"));
            if (attachPath != null) d.put("path", attachPath);
            for (Map.Entry<String, Object> entry : dict.entrySet()) {
                d.put(entry.getKey(), addAttachmentsPath(attachments, entry.getValue(), entry.getKey()));
            }

            return d;
        } else if (object instanceof List) {
            List<Object> list = (List<Object>) object;
            List<Object> l = new ArrayList<>(list.size());
            for (Object it : list) {
                l.add(addAttachmentsPath(attachments, it, null));
            }
            return l;
        }
        return object;
    }

    private void ensureInitialized() {
        if (mDatabase == null) throw new IllegalStateException("Database not initialized yet");
    }
}
